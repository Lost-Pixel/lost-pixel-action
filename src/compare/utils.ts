import { PNG } from 'pngjs';

export const resizeImage = (
  originalImage: PNG,
  width: number,
  height: number,
) => {
  const newImage = new PNG({
    width,
    height,
    fill: true,
    inputHasAlpha: true,
  });

  for (let x = 0; x < width; x++) {
    for (let y = 0; y < height; y++) {
      const index = ((width * y + x) << 2) + 3;
      newImage.data[index] = 64;
    }
  }

  PNG.bitblt(
    originalImage,
    newImage,
    0,
    0,
    originalImage.width,
    originalImage.height,
    0,
    0,
  );

  return newImage;
};
