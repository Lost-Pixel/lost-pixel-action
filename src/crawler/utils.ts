import http from 'node:http';
import handler from 'serve-handler';

export const launchStaticWebServer = async (basePath: string) => {
  const port = 3001;

  const server = http.createServer(async (request, response) => {
    return handler(request, response, {
      public: basePath.startsWith('file://') ? basePath.slice(7) : basePath,
    });
  });

  server.listen(port);

  return {
    server,
    port,
    url: `http://localhost:${port}`,
  };
};
