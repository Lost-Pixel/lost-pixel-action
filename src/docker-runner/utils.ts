import execa from 'execa';
import shell from 'shelljs';
import yargs from 'yargs';

import { hideBin } from 'yargs/helpers';

const isDockerInstalled = () => {
  if (!shell.which('docker')) {
    throw new Error('Docker is not installed');
  }
};

const isLostPixelImageDownloaded = async ({ version }: { version: string }) => {
  isDockerInstalled();
  const { exitCode, stdout, stderr } = await execa('docker', [
    'images',
    '-q',
    `lostpixel/lost-pixel:${version}`,
  ]);

  if (exitCode !== 0) {
    throw new Error(`Not successful docker operation, ${stderr}`);
  }

  return stdout.trim().length > 0;
};

const downloadImageIfNotExistent = async ({ version }: { version: string }) => {
  const doesImageExist = await isLostPixelImageDownloaded({
    version,
  });

  if (!doesImageExist) {
    await execa('docker', ['pull', `lostpixel/lost-pixel:v${version}`]);
  }
};

type ParsedYargs = {
  configDir: 'string';
};

export const executeDockerRun = async ({ version }: { version: string }) => {
  await downloadImageIfNotExistent({ version });

  // @ts-expect-error TBD
  const argv = yargs(hideBin(process.argv)).parse() as ParsedYargs;

  const args = [
    'run',
    '--rm',
    '-it',
    '-e WORKSPACE=/workspace',
    '-e DOCKER=1',
    `-e LOST_PIXEL_CONFIG_DIR=${argv.configDir}`,
    `lostpixel/lost-pixel:v${version}`,
  ];

  return execa('docker', args, { shell: true, stdio: 'inherit' });
};
