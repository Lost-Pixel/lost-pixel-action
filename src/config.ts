import { existsSync } from 'fs';
import { loadTSProjectConfigFile, log } from './utils';
import get from 'lodash.get';
import path from 'path';
import { BrowserContextOptions, Page } from 'playwright';

type BaseConfig = {
  /**
   * Browser to use: chromium, firefox, or webkit
   * @default 'chromium'
   */
  browser: 'chromium' | 'firefox' | 'webkit';

  /**
   * URL of the Lost Pixel API endpoint
   * @default 'https://app.lost-pixel.com/api/callback'
   */
  lostPixelUrl: string;

  /**
   * URL of the Storybook instance or local folder
   * @default 'storybook-static'
   */
  storybookUrl: string;

  /**
   * Path to the baseline image folder
   * @default '.lostpixel/baseline/'
   */
  imagePathBaseline: string;

  /**
   * Path to the current image folder
   * @default '.lostpixel/current/'
   */
  imagePathCurrent: string;

  /**
   * Path to the difference image folder
   * @default '.lostpixel/difference/'
   */
  imagePathDifference: string;

  /**
   * Number of concurrent shots to take
   * @default 5
   */
  shotConcurrency: number;

  /**
   * Number of concurrent screenshots to compare
   * @default 10
   */
  compareConcurrency: number;

  /**
   * Timeouts for various stages of the test
   */
  timeouts: {
    /**
     * Timeout for fetching stories from Storybook
     * @default 30_000
     */
    fetchStories?: number;

    /**
     * Timeout for loading the state of the page
     * @default 30_000
     */
    loadState?: number;

    /**
     * Timeout for waiting for network requests to finish
     * @default 30_000
     */
    networkRequests?: number;
  };

  /**
   * Time to wait before taking a screenshot
   * @default 1_000
   */
  waitBeforeScreenshot: number;

  /**
   * Time to wait for the first network request to start
   * @default 1_000
   */
  waitForFirstRequest: number;

  /**
   * Time to wait for the last network request to start
   * @default 1_000
   */
  waitForLastRequest: number;

  /**
   * Threshold for the difference between the baseline and current image
   *
   * Values between 0 and 1 are interpreted as percentage of the image size
   *
   * Values greater or equal to 1 are interpreted as pixel count.
   * @default 0
   */
  threshold: number;

  /**
   * Whether to set the GitHub status check on process start or not
   *
   * Setting this option to `true` makes only sense if the repository settings have pending status checks disabled
   * @default 'false'
   */
  setPendingStatusCheck: boolean;
};

type StoryLike = {
  id?: string;
  kind?: string;
  story?: string;
  parameters?: Record<string, unknown>;
};

export type ProjectConfig = {
  /**
   * Project ID
   */
  lostPixelProjectId: string;

  /**
   * CI build ID
   */
  ciBuildId: string;

  /**
   * CI build number
   */
  ciBuildNumber: string;

  /**
   * Git repository name (e.g. 'lost-pixel/lost-pixel-storybook')
   */
  repository: string;

  /**
   * Git branch name (e.g. 'refs/heads/main')
   */
  commitRef: string;

  /**
   * Git branch name (e.g. 'main')
   */
  commitRefName: string;

  /**
   * Git commit SHA (e.g. 'b9b8b9b9b9b9b9b9b9b9b9b9b9b9b9b9b9b9b9b9')
   */
  commitHash: string;

  /**
   * Git commit SHA (e.g. 'b9b8b9b9b9b9b9b9b9b9b9b9b9b9b9b9b9b9b9b9')
   */
  generateOnly: boolean;

  /**
   * S3 configuration
   */
  s3: {
    /**
     * S3 endpoint
     */
    endPoint: string;

    /**
     * S3 server port number
     */
    port?: number;

    /**
     * use SSL
     */
    ssl?: boolean;

    /**
     * S3 region
     */
    region?: string;

    /**
     * S3 access key
     */
    accessKey: string;

    /**
     * S3 secret key
     */
    secretKey: string;

    /**
     * S3 session token
     */
    sessionToken?: string;

    /**
     * S3 bucket name
     */
    bucketName: string;

    /**
     * S3 base URL
     */
    baseUrl?: string;
  };

  /**
   * File path to event.json file
   */
  eventFilePath?: string;

  /**
   * Global story filter
   */
  filterStory?: (input: StoryLike) => boolean;

  /**
   * File name generator for images
   */
  imageFilenameGenerator?: (input: StoryLike) => string;

  /**
   * Configure browser context options
   */
  configureBrowser?: (input: StoryLike) => BrowserContextOptions;

  /**
   * Configure page before screenshot
   */
  beforeScreenshot?: (page: Page, input: { id: string }) => Promise<void>;
};

const requiredConfigProps: Array<keyof FullConfig> = [
  'lostPixelProjectId',
  'ciBuildId',
  'ciBuildNumber',
  'repository',
  'commitRef',
  'commitRefName',
  'commitHash',
  's3',
];

const requiredS3ConfigProps: Array<keyof FullConfig['s3']> = [
  'endPoint',
  'accessKey',
  'secretKey',
  'bucketName',
];

export type FullConfig = BaseConfig & ProjectConfig;
export type CustomProjectConfig = Partial<BaseConfig> & ProjectConfig;

const defaultConfig: BaseConfig = {
  browser: 'chromium',
  lostPixelUrl: 'https://app.lost-pixel.com/api/callback',
  storybookUrl: 'storybook-static',
  imagePathBaseline: '.lostpixel/baseline/',
  imagePathCurrent: '.lostpixel/current/',
  imagePathDifference: '.lostpixel/difference/',
  shotConcurrency: 5,
  compareConcurrency: 10,
  timeouts: {
    fetchStories: 30_000,
    loadState: 30_000,
    networkRequests: 30_000,
  },
  waitBeforeScreenshot: 1_000,
  waitForFirstRequest: 1_000,
  waitForLastRequest: 1_000,
  threshold: 0,
  setPendingStatusCheck: false,
};

export let config: FullConfig;

const checkConfig = () => {
  const missingProps: string[] = [];

  const requiredProps = [
    ...requiredConfigProps,
    ...requiredS3ConfigProps.map((prop) => `s3.${prop}`),
  ];

  requiredProps.forEach((prop) => {
    if (!get(config, prop)) {
      missingProps.push(prop);
    }
  });

  if (missingProps.length > 0) {
    log(
      `Error: Missing required configuration properties: ${missingProps.join(
        ', ',
      )}`,
    );
    process.exit(1);
  }
};

const configFileNameBase = path.join(
  process.env.LOST_PIXEL_CONFIG_DIR || process.cwd(),
  'lostpixel.config',
);

const loadProjectConfig = async (): Promise<CustomProjectConfig> => {
  log('Loading project configuration...');
  log('Current working directory:', process.cwd());
  log('Defined configuration directory:', process.env.LOST_PIXEL_CONFIG_DIR);
  log('Looking for configuration file:', `${configFileNameBase}.(js|ts)`);

  if (existsSync(`${configFileNameBase}.js`)) {
    const projectConfig = require(`${configFileNameBase}.js`);
    return projectConfig;
  } else if (existsSync(`${configFileNameBase}.ts`)) {
    try {
      const imported = (await loadTSProjectConfigFile(
        `${configFileNameBase}.ts`,
      )) as CustomProjectConfig;
      return imported;
    } catch (error) {
      log(error);
      log('Failed to load TypeScript configuration file');
      process.exit(1);
    }
  }

  throw new Error("Couldn't find project config file 'lostpixel.config.js'");
};

export const configure = async (customProjectConfig?: CustomProjectConfig) => {
  if (customProjectConfig) {
    config = {
      ...defaultConfig,
      ...customProjectConfig,
    };

    return;
  }

  const projectConfig = await loadProjectConfig();

  config = {
    ...defaultConfig,
    ...projectConfig,
  };

  checkConfig();
};
