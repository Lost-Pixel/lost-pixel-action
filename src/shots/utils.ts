import { Page, Request } from 'playwright';
import { config } from '../config';

const checkIgnoreUrls = (url: string, ignoreUrls: string[]) => {
  for (const ignoreUrl of ignoreUrls) {
    if (url.includes(ignoreUrl)) {
      return true;
    }
  }

  return false;
};

export const waitForNetworkRequests = ({
  page,
  logger,
  timeout = config.timeouts.networkRequests,
  waitForFirstRequest = config.waitForFirstRequest,
  waitForLastRequest = config.waitForLastRequest,
  ignoreUrls = [],
}: {
  page: Page;
  logger: (message: string, ...rest: unknown[]) => void;
  timeout?: number;
  waitForFirstRequest?: number;
  waitForLastRequest?: number;
  ignoreUrls?: string[];
}) =>
  new Promise((resolve, reject) => {
    let requestCounter = 0;
    const requests = new Set<Request>();
    let lastRequestTimeoutId: NodeJS.Timeout;

    const timeoutId = setTimeout(() => {
      const pendingUrls = [...requests].map((request) => request.url());
      logger('Pending requests:', pendingUrls);

      cleanup();
      reject(new Error('Timeout'));
    }, timeout);

    const firstRequestTimeoutId = setTimeout(() => {
      cleanup();
      resolve(true);
    }, waitForFirstRequest);

    const onRequest = (request: Request) => {
      if (!checkIgnoreUrls(request.url(), ignoreUrls)) {
        clearTimeout(firstRequestTimeoutId);
        clearTimeout(lastRequestTimeoutId);
        requestCounter++;
        requests.add(request);
        logger(`+ ${request.url()}`);
      }
    };

    const onRequestFinished = async (request: Request) => {
      clearTimeout(lastRequestTimeoutId);

      if (!checkIgnoreUrls(request.url(), ignoreUrls)) {
        const failure = request.failure();
        const response = await request.response();

        requestCounter--;
        requests.delete(request);

        const statusText = failure
          ? failure.errorText
          : `${response?.status()} ${response?.statusText()}`;
        logger(`- ${request.url()} [${statusText}]`);
      }

      lastRequestTimeoutId = setTimeout(() => {
        if (requestCounter === 0) {
          cleanup();
          resolve(true);
        }
      }, waitForLastRequest);
    };

    function cleanup() {
      clearTimeout(timeoutId);
      clearTimeout(firstRequestTimeoutId);
      clearTimeout(lastRequestTimeoutId);
      page.removeListener('request', onRequest);
      page.removeListener('requestfinished', onRequestFinished);
      page.removeListener('requestfailed', onRequestFinished);
    }

    page.on('request', onRequest);
    page.on('requestfinished', onRequestFinished);
    page.on('requestfailed', onRequestFinished);
  });

export const resizeViewportToFullscreen = async ({ page }: { page: Page }) => {
  const height = await page.evaluate(
    () =>
      new Promise<number>((resolve) => {
        const body = document.body;
        const html = document.documentElement;

        resolve(
          Math.max(
            body.scrollHeight,
            body.offsetHeight,
            html.clientHeight,
            html.scrollHeight,
            html.offsetHeight,
          ),
        );
      }),
  );

  await page.setViewportSize({
    width: page.viewportSize()?.width || 800,
    height,
  });
};
