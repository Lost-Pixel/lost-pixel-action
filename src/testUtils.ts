import { CustomProjectConfig } from './config';

export const defaultTestConfig: CustomProjectConfig = {
  lostPixelProjectId: 'lostpixel-test',
  ciBuildId: '123',
  ciBuildNumber: '456',
  repository: 'lostpixel/lostpixel',
  commitRef: 'main',
  commitRefName: 'main',
  commitHash: 'a40a78068c77df941a488ef774a74ca781064d1c',
  s3: {
    endPoint: 's3.amazonaws.com',
    accessKey: '11111111111',
    secretKey: '22222222222',
    bucketName: 'lostpixel-demo',
  },
};
