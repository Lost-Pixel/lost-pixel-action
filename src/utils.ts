import {
  existsSync,
  mkdirSync,
  readdirSync,
  unlinkSync,
  writeFileSync,
  readFileSync,
} from 'node:fs';
import * as crypto from 'node:crypto';
import type { Buffer } from 'node:buffer';
import { normalize, join } from 'node:path';
import { PostHog } from 'posthog-node';
import { v4 as uuid } from 'uuid';
import { type BrowserType, chromium, firefox, webkit } from 'playwright-core';
import yargs from 'yargs';
import { hideBin } from 'yargs/helpers';
import { config } from './config';
import { log } from './log';
import type { ShotItem } from './types';

export type ParsedYargs = {
  _: ['update', 'meta', 'docker', 'local'];
  m: 'update';
};

type FilenameWithPath = {
  name: string;
  path: string;
};

type FilenameWithAllPaths = {
  name: string;
  path: string;
  pathCurrent?: string;
};

type Files = {
  baseline: FilenameWithPath[];
  current: FilenameWithPath[];
  difference: FilenameWithPath[];
};

export type Changes = {
  difference: FilenameWithAllPaths[];
  deletion: FilenameWithAllPaths[];
  addition: FilenameWithAllPaths[];
};

const POST_HOG_API_KEY = 'phc_RDNnzvANh1mNm9JKogF9UunG3Ky02YCxWP9gXScKShk';

export const isUpdateMode = (): boolean => {
  // @ts-expect-error TBD
  const args = yargs(hideBin(process.argv)).parse() as ParsedYargs;

  return (
    args._.includes('update') ||
    args.m === 'update' ||
    process.env.LOST_PIXEL_MODE === 'update'
  );
};

export const isDockerMode = (): boolean => {
  // @ts-expect-error TBD
  const args = yargs(hideBin(process.argv)).parse() as ParsedYargs;

  return args._.includes('docker') || process.env.LOST_PIXEL_DOCKER === 'true';
};

export const shallGenerateMeta = (): boolean => {
  // @ts-expect-error TBD
  const args = yargs(hideBin(process.argv)).parse() as ParsedYargs;

  return (
    args._.includes('meta') || process.env.LOST_PIXEL_GENERATE_META === 'true'
  );
};

export const getChanges = (files: Files): Changes => {
  return {
    difference: files.difference
      .map((file) => ({
        ...file,
        pathCurrent: files.current.find(({ name }) => name === file.name)?.path, // Keep track of custom shots path
      }))
      .sort((a, b) => a.name.localeCompare(b.name)),
    deletion: files.baseline
      .filter(
        (file1) => !files.current.some((file2) => file1.name === file2.name),
      )
      .sort((a, b) => a.name.localeCompare(b.name)),
    addition: files.current
      .filter(
        (file1) => !files.baseline.some((file2) => file1.name === file2.name),
      )
      .sort((a, b) => a.name.localeCompare(b.name)),
  };
};

type ExtendFileName = {
  fileName: string;
  extension: 'after' | 'before' | 'difference';
};

export const extendFileName = ({ fileName, extension }: ExtendFileName) => {
  const parts = fileName.split('.').filter((part) => part !== '');
  const extensionIndex = parts.length - 1;

  if (parts.length === 1) {
    return `${extension}.${parts[0]}`;
  }

  if (parts.length === 0) {
    return extension;
  }

  parts[extensionIndex] = `${extension}.${parts[extensionIndex]}`;

  return parts.join('.');
};

export const createShotsFolders = () => {
  const paths = [
    config.imagePathBaseline,
    config.imagePathCurrent,
    config.imagePathDifference,
  ];

  for (const path of paths) {
    if (!existsSync(path)) {
      mkdirSync(path, { recursive: true });
    }
  }

  const ignoreFile = normalize(
    join(config.imagePathBaseline, '..', '.gitignore'),
  );

  if (!existsSync(ignoreFile)) {
    writeFileSync(ignoreFile, 'current\ndifference\n');
  }
};

export const sleep = async (ms: number) =>
  new Promise((resolve) => {
    setTimeout(resolve, ms);
  });

export const removeFilesInFolder = (path: string, excludePaths?: string[]) => {
  const files = readdirSync(path);

  const filesPathsIgnoringExclude = files
    .map((file) => join(path, file))
    .filter((filePath) => !excludePaths?.includes(filePath));

  log.process(
    'info',
    'general',
    `Removing ${filesPathsIgnoringExclude.length} files from ${path}`,
  );

  for (const filePath of filesPathsIgnoringExclude) {
    unlinkSync(filePath);
  }
};

export const getBrowser = (): BrowserType => {
  switch (config.browser) {
    case 'chromium': {
      return chromium;
    }

    case 'firefox': {
      return firefox;
    }

    case 'webkit': {
      return webkit;
    }

    default: {
      return chromium;
    }
  }
};

export const getVersion = (): string | void => {
  try {
    // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment, @typescript-eslint/no-require-imports, @typescript-eslint/no-var-requires
    const packageJson: { version: string } = require('../package.json');

    return packageJson.version;
  } catch {}
};

const fileNameWithoutExtension = (fileName: string): string => {
  return fileName.split('.').slice(0, -1).join('.');
};

export const readDirIntoShotItems = (path: string): ShotItem[] => {
  const files = readdirSync(path);

  return files
    .filter((name) => name.endsWith('.png'))
    .map((fileNameWithExt): ShotItem => {
      const fileName = fileNameWithoutExtension(fileNameWithExt);

      return {
        id: fileName,
        shotName: fileName,
        shotMode: 'custom',
        filePathBaseline: join(config.imagePathBaseline, fileNameWithExt),
        filePathCurrent: join(path, fileNameWithExt),
        filePathDifference: join(config.imagePathDifference, fileNameWithExt),
        url: fileName,
        // TODO: custom shots take thresholds only from config - not possible to source configs from individual story
        threshold: config.threshold,
      };
    });
};

const sendTelemetryData = async (properties: {
  runDuration?: number;
  shotsNumber?: number;
  error?: unknown;
}) => {
  const client = new PostHog(POST_HOG_API_KEY);
  const id: string = uuid();

  try {
    log.process('info', 'general', 'Sending anonymized telemetry data.');

    const version = getVersion() as string;
    const modes = [];

    if (config.storybookShots) modes.push('storybook');

    if (config.ladleShots) modes.push('ladle');

    if (config.pageShots) modes.push('pages');

    if (config.customShots) modes.push('custom');

    if (properties.error) {
      client.capture({
        distinctId: id,
        event: 'lost-pixel-error',
        properties: { ...properties },
      });
    } else {
      client.capture({
        distinctId: id,
        event: 'lost-pixel-run',
        properties: { ...properties, version, modes },
      });
    }

    await client.shutdownAsync();
  } catch (error: unknown) {
    log.process('error', 'general', 'Error when sending telemetry data', error);
  }
};

export const parseHrtimeToSeconds = (hrtime: [number, number]) => {
  const seconds = (hrtime[0] + hrtime[1] / 1e9).toFixed(3);

  return seconds;
};

export const exitProcess = async (properties: {
  runDuration?: number;
  shotsNumber?: number;
  error?: unknown;
  exitCode?: 0 | 1;
}) => {
  if (process.env.LOST_PIXEL_DISABLE_TELEMETRY === '1') {
    process.exit(properties.exitCode ?? 1);
  } else {
    await sendTelemetryData(properties).finally(() => {
      process.exit(properties.exitCode ?? 1);
    });
  }
};

const hashBuffer = (buffer: Buffer): string => {
  const hashSum = crypto.createHash('sha256');

  hashSum.update(buffer);

  return hashSum.digest('hex');
};

export const hashFile = (filePath: string): string => {
  const file = readFileSync(filePath);

  return hashBuffer(file);
};
