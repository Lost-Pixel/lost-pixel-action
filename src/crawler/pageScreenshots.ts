import path from 'node:path';
import axios, { isAxiosError } from 'axios';
import { z } from 'zod';
import type { BrowserType } from 'playwright-core';
import fs from 'fs-extra';
import { log } from '../log';
import { type Mask, type PageScreenshotParameter, config } from '../config';
import type { ShotItem } from '../types';
import { selectBreakpoints, generateLabel } from '../shots/utils';

const generateBrowserConfig = (page: PageScreenshotParameter) => {
  const browserConfig = config.configureBrowser?.({
    ...page,
    shotMode: 'page',
  });

  if (page.viewport && browserConfig) {
    browserConfig.viewport = browserConfig.viewport ?? {
      width: 1280,
      height: 720,
    };
    browserConfig.viewport = {
      ...browserConfig.viewport,
      ...page.viewport,
    };
  }

  return browserConfig;
};

export const generatePageShotItems = (
  pages: PageScreenshotParameter[],
  baseUrl: string,
  mask?: Mask[],
  modeBreakpoints?: number[],
  browser?: BrowserType,
): ShotItem[] => {
  const names = pages.map((page) => page.name);
  const uniqueNames = new Set(names);

  if (names.length !== uniqueNames.size) {
    throw new Error('Error: Page names must be unique');
  }

  return pages.flatMap((page): ShotItem[] => {
    const shotName =
      config.shotNameGenerator?.({ ...page, shotMode: 'page' }) ?? page.name;
    let label = generateLabel({ browser });
    let fileNameWithExt = `${shotName}${label}.png`;

    const baseShotItem: ShotItem = {
      shotMode: 'page',
      id: `${shotName}${label}`,
      shotName: `${shotName}${label}`,
      url: path.join(baseUrl, page.path),
      filePathBaseline: path.join(config.imagePathBaseline, fileNameWithExt),
      filePathCurrent: path.join(config.imagePathCurrent, fileNameWithExt),
      filePathDifference: path.join(
        config.imagePathDifference,
        fileNameWithExt,
      ),
      browserConfig: generateBrowserConfig(page),
      threshold: page.threshold ?? config.threshold,
      waitBeforeScreenshot:
        page.waitBeforeScreenshot ?? config.waitBeforeScreenshot,
      mask: [...(mask ?? []), ...(page.mask ?? [])],
    };

    const breakpoints = selectBreakpoints(
      config.breakpoints,
      modeBreakpoints,
      page.breakpoints,
    );

    if (!breakpoints || breakpoints.length === 0) {
      return [baseShotItem];
    }

    return breakpoints.map((breakpoint) => {
      label = generateLabel({ breakpoint, browser });
      fileNameWithExt = `${shotName}${label}.png`;

      return {
        ...baseShotItem,
        id: `${shotName}${label}`,
        shotName: `${shotName}${label}`,
        breakpoint,
        breakpointGroup: page.name,
        url: path.join(baseUrl, page.path),
        filePathBaseline: path.join(config.imagePathBaseline, fileNameWithExt),
        filePathCurrent: path.join(config.imagePathCurrent, fileNameWithExt),
        filePathDifference: path.join(
          config.imagePathDifference,
          fileNameWithExt,
        ),
        viewport: { width: breakpoint },
        browserConfig: generateBrowserConfig({
          ...page,
          viewport: { width: breakpoint },
        }),
      };
    });
  });
};

// Helper function to check if a string is a valid URL
const isValidHttpUrl = (string: string) => {
  let url;

  try {
    url = new URL(string);
  } catch {
    return false;
  }

  return url.protocol === 'http:' || url.protocol === 'https:';
};

export const getPagesFromExternalLoader = async () => {
  try {
    if (!config.pageShots?.pagesJsonUrl) {
      return [];
    }

    log.process(
      'info',
      'general',
      `‚è¨ Loading pages from ${config.pageShots.pagesJsonUrl}`,
    );

    let pages;

    // Check if the pagesJsonUrl is a valid URL or a local file path
    if (isValidHttpUrl(config.pageShots.pagesJsonUrl)) {
      log.process('info', 'general', `üï∏Ô∏è Trying to fetch from URL`);
      const response = await axios.get<PageScreenshotParameter[]>(
        config.pageShots.pagesJsonUrl,
      );

      pages = response.data;
    } else {
      // Read the file from the local filesystem
      log.process('info', 'general', `‚è¨ Trying to fetch from local file`);
      const fileContents = await fs.readFile(
        config.pageShots.pagesJsonUrl,
        'utf8',
      );

      pages = JSON.parse(fileContents) as PageScreenshotParameter[];
    }

    // Validation logic remains the same
    const pagesArraySchema = z.array(
      z.object({
        path: z.string(),
        name: z.string(),
        waitBeforeScreenshot: z.number().optional(),
        threshold: z.number().optional(),
        mask: z
          .array(
            z.object({
              selector: z.string(),
            }),
          )
          .optional(),
        viewport: z
          .object({
            width: z.string(),
            height: z.string(),
          })
          .optional(),
      }),
    );

    const validatePages = pagesArraySchema.safeParse(pages);

    if (validatePages.success) {
      log.process(
        'info',
        'general',
        `‚úÖ Successfully validated pages structure & loaded ${pages.length} pages from JSON file.`,
      );

      return pages;
    }

    log.process(
      'error',
      'general',
      '‚ùå Error validating the loaded pages structure',
    );
    log.process('error', 'general', validatePages.error);

    return [];
  } catch (error: unknown) {
    if (isAxiosError(error) || error instanceof Error) {
      log.process(
        'error',
        'network',
        `‚ùå Error when fetching data: ${error.message}`,
      );
    }

    return [];
  }
};
