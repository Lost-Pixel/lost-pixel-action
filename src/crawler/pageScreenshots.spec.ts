import { createShotsFolders } from '../utils';
import { type PageScreenshotParameter, configure } from '../config';
import { defaultTestConfig } from '../testUtils';
import { generatePageShotItems } from './pageScreenshots';

beforeAll(async () => {
  await configure({
    ...defaultTestConfig,
    timeouts: {
      fetchStories: 2000,
    },
  });

  createShotsFolders();
  process.env.FETCH_STORIES_TIMEOUT = '2000';
});

describe(generatePageShotItems, () => {
  it('should generate shot items for pages without breakpoints', () => {
    const baseUrl = 'https://example.com';

    const pages: PageScreenshotParameter[] = [
      {
        name: 'home',
        path: '/',
      },
      {
        name: 'about',
        path: '/about',
      },
    ];

    const shotItems = generatePageShotItems(pages, baseUrl);

    expect(shotItems).toMatchSnapshot('PagesWithoutBreakpoints');
  });

  it('should generate shot items for pages with breakpoints', () => {
    const baseUrl = 'https://example.com';

    const pages: PageScreenshotParameter[] = [
      {
        name: 'home',
        path: '/',
        breakpoints: [480, 768],
      },
      {
        name: 'about',
        path: '/about',
        breakpoints: [480, 768],
      },
    ];

    const shotItems = generatePageShotItems(pages, baseUrl);

    expect(shotItems).toMatchSnapshot('PagesWithBreakpoints');
  });
});
