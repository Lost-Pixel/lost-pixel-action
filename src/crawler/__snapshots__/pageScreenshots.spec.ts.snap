// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generatePageShotItems should generate shot items for pages with breakpoints: PagesWithBreakpoints 1`] = `
[
  {
    "breakpoint": 480,
    "breakpointGroup": "home",
    "browserConfig": undefined,
    "filePathBaseline": ".lostpixel/baseline/home__[w480px].png",
    "filePathCurrent": ".lostpixel/current/home__[w480px].png",
    "filePathDifference": ".lostpixel/difference/home__[w480px].png",
    "id": "home[__[w480px]]",
    "mask": [],
    "shotMode": "page",
    "shotName": "home",
    "threshold": 0,
    "url": "https:/example.com/",
    "viewport": {
      "width": 480,
    },
    "waitBeforeScreenshot": 1000,
  },
  {
    "breakpoint": 768,
    "breakpointGroup": "home",
    "browserConfig": undefined,
    "filePathBaseline": ".lostpixel/baseline/home__[w768px].png",
    "filePathCurrent": ".lostpixel/current/home__[w768px].png",
    "filePathDifference": ".lostpixel/difference/home__[w768px].png",
    "id": "home[__[w768px]]",
    "mask": [],
    "shotMode": "page",
    "shotName": "home",
    "threshold": 0,
    "url": "https:/example.com/",
    "viewport": {
      "width": 768,
    },
    "waitBeforeScreenshot": 1000,
  },
  {
    "breakpoint": 480,
    "breakpointGroup": "about",
    "browserConfig": undefined,
    "filePathBaseline": ".lostpixel/baseline/about__[w480px].png",
    "filePathCurrent": ".lostpixel/current/about__[w480px].png",
    "filePathDifference": ".lostpixel/difference/about__[w480px].png",
    "id": "about[__[w480px]]",
    "mask": [],
    "shotMode": "page",
    "shotName": "about",
    "threshold": 0,
    "url": "https:/example.com/about",
    "viewport": {
      "width": 480,
    },
    "waitBeforeScreenshot": 1000,
  },
  {
    "breakpoint": 768,
    "breakpointGroup": "about",
    "browserConfig": undefined,
    "filePathBaseline": ".lostpixel/baseline/about__[w768px].png",
    "filePathCurrent": ".lostpixel/current/about__[w768px].png",
    "filePathDifference": ".lostpixel/difference/about__[w768px].png",
    "id": "about[__[w768px]]",
    "mask": [],
    "shotMode": "page",
    "shotName": "about",
    "threshold": 0,
    "url": "https:/example.com/about",
    "viewport": {
      "width": 768,
    },
    "waitBeforeScreenshot": 1000,
  },
]
`;

exports[`generatePageShotItems should generate shot items for pages without breakpoints: PagesWithoutBreakpoints 1`] = `
[
  {
    "browserConfig": undefined,
    "filePathBaseline": ".lostpixel/baseline/home.png",
    "filePathCurrent": ".lostpixel/current/home.png",
    "filePathDifference": ".lostpixel/difference/home.png",
    "id": "home",
    "mask": [],
    "shotMode": "page",
    "shotName": "home",
    "threshold": 0,
    "url": "https:/example.com/",
    "waitBeforeScreenshot": 1000,
  },
  {
    "browserConfig": undefined,
    "filePathBaseline": ".lostpixel/baseline/about.png",
    "filePathCurrent": ".lostpixel/current/about.png",
    "filePathDifference": ".lostpixel/difference/about.png",
    "id": "about",
    "mask": [],
    "shotMode": "page",
    "shotName": "about",
    "threshold": 0,
    "url": "https:/example.com/about",
    "waitBeforeScreenshot": 1000,
  },
]
`;
